### PR TITLE
Add custom fields to TimeEntry

### PR DESCRIPTION
**Pull request by christau**
In case you find it useful too.

### CODE REVIEW

**Comment by alexeyOnGitHub** [comment]
how did you test this? I am hesitant to merge any code changes that do not have corresponding tests.

**Comment by christau** [comment]
I wrote a litte java app for myself, that automatically creates invoices from time entries in redmine. But the time entries in my redmine instance had some cutom fields which i also needed in this java app. Since redmine java api had a clean interface and everything was almost there, i decided to add a few lines for accessing those custom fields. And as it worked for me i thought maybe it's a good idea to have it in the original project, so i don't have to use the custom built jar anymore and others can take advantage of these features from redmine.
Long story short. I just don't know how to write a test for this.

File: src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java [hunk 1]
```diff
@@ -139,13 +143,25 @@ public String getUserName() {
     public void setUserName(String userName) {
         storage.set(USER_NAME, userName);
     }
+    
+    public CustomField getCustomField(String name) {
+    		return 	((List<CustomField>)storage.get(CUSTOM_FIELDS)).stream().filter(a -> a.getName().equals(name)).findFirst().get();
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java]: I wonder if there was a particular reason to use a non-generified Property<List> for CUSTOM_FIELDS field?
see 
`public final static Property<Set<CustomField>> CUSTOM_FIELDS = (Property<Set<CustomField>>) new Property(Set.class, "customFields");`
in Issue class for comparison.
this allows to avoid type casting later, like this:
` for (CustomField customField : storage.get(CUSTOM_FIELDS)) {`

File: src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java [hunk 2]
```diff
@@ -31,6 +34,7 @@
     public final static Property<Date> SPENT_ON = new Property<>(Date.class, "spentOn");
     public final static Property<Date> CREATED_ON = new Property<>(Date.class, "createdOn");
     public final static Property<Date> UPDATED_ON = new Property<>(Date.class, "updatedOn");
+	public static final Property<List> CUSTOM_FIELDS = new Property<>(List.class, "custom_fields");
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java]: how about using Set instead of List? do you expect duplicate entries there? if yes, then how would they be resolved?

File: src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java (hunk 1 above)
Comment (alexeyOnGitHub) [reply src/main/java/com/taskadapter/redmineapi/bean/TimeEntry.java]: findFirst().get() will produce an exception if nothing is found. this is inconsistent with the way custom fields are resolved in Issue class.
a better way would be to return Optional itself. but at least need to be consistent with the other classes.

File: src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONParser.java [hunk 3]
```diff
@@ -357,6 +359,19 @@ public static CustomField parseCustomField(JSONObject content)
...
+		for(int i = 0; i < content.length(); i++)
+		{
+			JSONObject object = content.getJSONObject(i);
+			results.add(CustomFieldFactory.create(JsonInput.getInt(object, "id"),
+					JsonInput.getStringOrEmpty(object, "name"),
+					JsonInput.getStringOrEmpty(object, "value")));
+		}
+			
+		return results;
+	}
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONParser.java]: did you see how Project element is parsed? 
`  result.addCustomFields(JsonInput.getListOrEmpty(content, "custom_fields", RedmineJSONParser::parseCustomField));`

was there a reason why the same code did not work for you and you had to implement `parseCustomFields` method from scratch?

File: src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONParser.java [hunk 4]
```diff
@@ -155,6 +155,8 @@ public static TimeEntry parseTimeEntry(JSONObject object)
 		result.setSpentOn(getDateOrNull(object, "spent_on"));
 		result.setCreatedOn(getDateOrNull(object, "created_on"));
 		result.setUpdatedOn(getDateOrNull(object, "updated_on"));
+		final JSONArray custom_fields = JsonInput.getArrayOrNull(object, "custom_fields");
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONParser.java]: getArrayOrNull name implies that null will be returned if nothing is found. then later the result of this method is passed to  parseCustomFields()   where it is referenced directly, thus allowing NullPointerException to happen.

**Comment by alexeyOnGitHub** [comment]
side note - I looked at how custom fields are done for time tracking in Redmine and this just does not make sense to me. I cannot figure out what this is supposed to mean -

**Comment by alexeyOnGitHub** [comment]
image

**Comment by alexeyOnGitHub** [comment]
what are these values in custom fields in "Activities" ? and why don't I see custom fields when I click "log time" on an issue?

File: src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONBuilder.java [hunk 5]
```diff
@@ -75,6 +75,7 @@ static void writeTimeEntry(JSONWriter writer, TimeEntry timeEntry)
 		addIfSetShort2(writer, "spent_on", storage, TimeEntry.SPENT_ON);
 		addIfSetFullDate(writer, "created_on", storage, TimeEntry.SPENT_ON);
 		addIfSetFullDate(writer, "updated_on", storage, TimeEntry.SPENT_ON);
+		addIfSet(writer, "custom_fields", storage, TimeEntry.CUSTOM_FIELDS);
```
Comment (alexeyOnGitHub) [inline src/main/java/com/taskadapter/redmineapi/internal/RedmineJSONBuilder.java]: this gives Redmine Server error 500 in my simple test.
I changed this to 
`writeCustomFields(writer, timeEntry.getCustomFields());`
- again, similarly to how it is done for other elements, like Project.

**Comment by alexeyOnGitHub** [comment]
I made some changed and merged the end result to "master". let me know if this version works for you

**Comment by christau** [comment]
Thanks very much for the review and the merge! It works for me.
